### PR TITLE
*: refine join result generator to return MaxChunkSize chunk

### PR DESCRIPTION
**Pull request by XuHuaiyu**
Refine join result generator to return at most MaxChunkSize chunk every time.

As the diagram shows,
this commit only cost 1/4 memory usage when execute a  Cartesian sql(as the 18:26~18:38 part shows),
and save almost 47% memory usage on my machine when run tpch5.

PTAL @zz-jason @coocood

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @coocood

File: executor/index_lookup_join.go [hunk 1]
```diff
@@ -55,6 +55,8 @@ type IndexLookUpJoin struct {
 	task             *lookUpJoinTask
 	joinResult       *chunk.Chunk
 	joinResultCursor int
+	innerIter        chunk.Iterator
+	innerRowBuffer   chunk.Row
```
Comment (zz-jason) [inline executor/index_lookup_join.go]: s/innerRowBuffer/innerRow/

File: executor/join.go [hunk 2]
```diff
@@ -667,40 +665,52 @@ func (e *HashJoinExec) joinOuterRow(workerID int, outerRow Row, resultBuffer *ex
...
-
-	err = e.resultGenerators[workerID].emitToChunk(outerRow, chunk.NewIterator4Slice(innerRows), chk)
-	if err != nil {
-		return errors.Trace(err)
+	iter := chunk.NewIterator4Slice(innerRows)
+	for inner := iter.Begin(); inner != iter.End(); {
+		inner, err = e.resultGenerators[workerID].emitToChunk(outerRow, iter, inner, joinResult.chk)
+		if err != nil {
+			joinResult.err = errors.Trace(err)
+			break
```
Comment (zz-jason) [inline executor/join.go]: `return false, joinResult`

File: executor/join.go [hunk 3]
```diff
@@ -667,40 +665,52 @@ func (e *HashJoinExec) joinOuterRow(workerID int, outerRow Row, resultBuffer *ex
...
+			ok := true
+			e.joinResultCh <- joinResult
+			ok, joinResult = e.getNewJoinResult(workerID)
+			if !ok {
+				return false, joinResult
+			}
+		}
 	}
-	return nil
+	return joinResult.err == nil, joinResult
```
Comment (zz-jason) [inline executor/join.go]: `return true, joinResult`

File: executor/join.go [hunk 4]
```diff
@@ -716,37 +726,32 @@ func (e *HashJoinExec) getNewJoinResult(workerID int) (bool, *hashjoinWorkerResu
...
+			_, err = e.resultGenerators[workerID].emitToChunk(outerChk.GetRow(i), nil, chunk.Row{}, joinResult.chk)
+			if err != nil {
+				joinResult.err = errors.Trace(err)
+				return false, joinResult
+			}
 		} else { // process matched outer rows
-			err = e.joinMatchedOuterRow2Chunk(workerID, outerChk.GetRow(i), joinResultChkBuffer)
+			ok, joinResult = e.joinMatchedOuterRow2Chunk(workerID, outerChk.GetRow(i), joinResult)
+			if !ok {
+				return ok, joinResult
```
Comment (zz-jason) [inline executor/join.go]: `return false, joinResult`

File: executor/join.go [hunk 5]
```diff
@@ -839,7 +844,9 @@ type NestedLoopApplyExec struct {
 	innerList        *chunk.List
 	innerChunk       *chunk.Chunk
 	innerSelected    []bool
-	resultChunk      *chunk.Chunk
+	innerIter        chunk.Iterator
+	innerRowBuffer   chunk.Row
+	outerRowBuffer   *chunk.Row
```
Comment (zz-jason) [inline executor/join.go]: s/innerRowBuffer/innerRow/
s/outerRowBuffer/outerRow/

File: executor/join_result_generators.go [hunk 6]
```diff
@@ -48,7 +48,7 @@ type joinResultGenerator interface {
 	emit(outer Row, inners []Row, resultBuffer []Row) ([]Row, error)
 
 	// emitToChunk takes the same operation as emit, but the joined rows is appended to a Chunk instead of a result buffer.
-	emitToChunk(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) error
+	emitToChunk(outer chunk.Row, inners chunk.Iterator, inner chunk.Row, chk *chunk.Chunk) (chunk.Row, error)
```
Comment (zz-jason) [inline executor/join_result_generators.go]: s/inner/innerBegin/

File: executor/join_result_generators.go [hunk 7]
```diff
@@ -48,7 +48,7 @@ type joinResultGenerator interface {
 	emit(outer Row, inners []Row, resultBuffer []Row) ([]Row, error)
 
 	// emitToChunk takes the same operation as emit, but the joined rows is appended to a Chunk instead of a result buffer.
```
Comment (zz-jason) [inline executor/join_result_generators.go]: update comment

File: executor/merge_join.go [hunk 8]
```diff
@@ -52,9 +51,12 @@ type MergeJoinExec struct {
 	// for chunk execution.
 	compareFuncs   []chunk.CompareFunc
 	outerIdx       int
-	resultChunk    *chunk.Chunk
 	outerChunkRows []chunk.Row
 	innerChunkRows []chunk.Row
+	outerIter4Row  chunk.Iterator
+	innerIter4Row  chunk.Iterator
+	innerRowBuffer chunk.Row
+	outerRowBuffer chunk.Row
```
Comment (zz-jason) [inline executor/merge_join.go]: s/innerRowBuffer/innerRow/
s/outerRowBuffer/outerRow/

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @coocood

File: executor/merge_join.go [hunk 9]
```diff
@@ -470,80 +477,85 @@ func (e *MergeJoinExec) Next(goCtx goctx.Context) (Row, error) {
...
-			return true, nil
+		if err = e.fetchNextOuterRows(); err != nil {
+			return false, errors.Trace(err)
 		}
+
+	}
+	return false, nil
+}
+
+func (e *MergeJoinExec) fetchInnerRows() (err error) {
```
Comment (zz-jason) [inline executor/merge_join.go]: s/fetchInnerRows/fetchNextInnerRows/

File: util/chunk/iterator.go [hunk 10]
```diff
@@ -95,21 +118,35 @@ func (it *Iterator4Chunk) Begin() Row {
...
 	}
 	row := it.chk.GetRow(it.cursor)
 	it.cursor++
 	return row
 }
 
-// End implements the Iterator interface
+// Current implements the Iterator interface.
+func (it *Iterator4Chunk) Current() Row {
+	if it.chk.NumRows() == 0 || it.cursor > it.chk.NumRows() {
```
Comment (zz-jason) [inline util/chunk/iterator.go]: s/it.cursor > it.chk.NumRows()/it.cursor > it.Len()/

File: util/chunk/iterator.go [hunk 11]
```diff
@@ -95,21 +118,35 @@ func (it *Iterator4Chunk) Begin() Row {
...
+}
+
+// End implements the Iterator interface.
 func (it *Iterator4Chunk) End() Row {
 	return Row{}
 }
 
+// ReachEnd implements the Iterator interface.
+func (it *Iterator4Chunk) ReachEnd() {
+	it.cursor = it.chk.NumRows()
```
Comment (zz-jason) [inline util/chunk/iterator.go]: s/it.chk.NumRows()/it.Len()/

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @coocood

File: util/chunk/iterator.go [hunk 12]
```diff
@@ -58,19 +64,36 @@ func (it *iterator4Slice) Begin() Row {
 	return it.rows[0]
 }
 
+// Next implements the Iterator interface.
 func (it *iterator4Slice) Next() Row {
-	if it.cursor == len(it.rows) {
+	if it.cursor >= len(it.rows) {
```
Comment (zz-jason) [inline util/chunk/iterator.go]: s/len(it.rows)/it.Len()/

File: util/chunk/iterator.go [hunk 13]
```diff
@@ -58,19 +64,36 @@ func (it *iterator4Slice) Begin() Row {
...
+}
+
+// End implements the Iterator interface.
 func (it *iterator4Slice) End() Row {
 	return Row{}
 }
 
+// ReachEnd implements the Iterator interface.
+func (it *iterator4Slice) ReachEnd() {
+	it.cursor = len(it.rows) + 1
```
Comment (zz-jason) [inline util/chunk/iterator.go]: ditto

File: util/chunk/iterator.go [hunk 14]
```diff
@@ -95,21 +118,35 @@ func (it *Iterator4Chunk) Begin() Row {
 	return it.chk.GetRow(0)
 }
 
-// Next implements the Iterator interface
+// Next implements the Iterator interface.
 func (it *Iterator4Chunk) Next() Row {
-	if it.cursor == it.chk.NumRows() {
+	if it.cursor >= it.chk.NumRows() {
```
Comment (zz-jason) [inline util/chunk/iterator.go]: s/it.chk.NumRows()/it.Len()/

File: util/chunk/iterator.go [hunk 15]
```diff
@@ -183,19 +244,36 @@ func (it *iterator4RowPtr) Begin() Row {
 	return it.li.GetRow(it.ptrs[0])
 }
 
+// Next implements the Iterator interface.
 func (it *iterator4RowPtr) Next() Row {
-	if it.cursor == len(it.ptrs) {
+	if it.cursor >= len(it.ptrs) {
```
Comment (zz-jason) [inline util/chunk/iterator.go]: s/len(it.ptrs)/it.Len()/

**Comment by XuHuaiyu** [comment]
PTAL @coocood

File: util/chunk/iterator.go [hunk 16]
```diff
@@ -175,27 +235,45 @@ type iterator4RowPtr struct {
...
 		return it.End()
 	}
 	row := it.li.GetRow(it.ptrs[it.cursor])
 	it.cursor++
 	return row
 }
 
+// Current implements the Iterator interface.
+func (it *iterator4RowPtr) Current() Row {
+	if len := it.Len(); len == 0 || it.cursor > len {
```
Comment (coocood) [inline util/chunk/iterator.go]: How about
```
if it.cursor >= it.Len() {
    return it.End()
}
```

File: util/chunk/iterator.go [hunk 17]
```diff
@@ -175,27 +235,45 @@ type iterator4RowPtr struct {
...
+		return it.End()
+	}
```
Comment (coocood) [inline util/chunk/iterator.go]: Can be simplified to
```
if it.cursor  == 0 || it.cursor > it.Len() {
```

File: executor/join_result_generators.go [hunk 18]
```diff
@@ -491,22 +499,26 @@ func (outputer *leftOuterJoinResultGenerator) emitToChunk(outer chunk.Row, inner
 	if len(outputer.filter) == 0 {
 		chkForJoin = chk
 	}
-	for inner := inners.Begin(); inner != inners.End(); inner = inners.Next() {
-		outputer.makeJoinRowToChunk(chkForJoin, outer, inner)
-	}
-	if len(outputer.filter) == 0 {
-		return nil
-	}
+	for inner, numToAppend := inners.Current(), outputer.maxChunkSize-chk.NumRows(); inner != inners.End() && numToAppend > 0; numToAppend = outputer.maxChunkSize - chk.NumRows() {
```
Comment (coocood) [inline executor/join_result_generators.go]: This line is too long, how about declaring the two variable before the for loop?

File: executor/join_result_generators.go [hunk 19]
```diff
@@ -613,23 +629,29 @@ func (outputer *innerJoinResultGenerator) emitToChunk(outer chunk.Row, inners ch
 	if len(outputer.filter) == 0 {
 		chkForJoin = chk
 	}
-	if outputer.outerIsRight {
-		for inner := inners.Begin(); inner != inners.End(); inner = inners.Next() {
-			outputer.makeJoinRowToChunk(chkForJoin, inner, outer)
+	for inner, numToAppend := inners.Current(), outputer.maxChunkSize-chk.NumRows(); inner != inners.End() && numToAppend > 0; numToAppend = outputer.maxChunkSize - chk.NumRows() {
```
Comment (coocood) [inline executor/join_result_generators.go]: ditto

File: util/chunk/iterator_test.go [hunk 20]
```diff
@@ -43,16 +43,55 @@ func (s *testChunkSuite) TestIterator(c *check.C) {
 
 	it := NewIterator4Slice(rows)
 	checkIterator(c, it, expected)
+	for i := 0; i > 5; i++ {
```
Comment (lamxTyler) [inline util/chunk/iterator_test.go]: Should be `i < 5`?

File: util/chunk/iterator_test.go [hunk 21]
```diff
@@ -43,16 +43,55 @@ func (s *testChunkSuite) TestIterator(c *check.C) {
...
+		c.Assert(it.Current(), check.Equals, chk.GetRow(i))
+		it.Next()
+	}
+	it.ReachEnd()
+	c.Assert(it.Current(), check.Equals, it.End())
+	c.Assert(it.Begin(), check.Equals, chk.GetRow(0))
+
 	it = NewIterator4List(li)
 	checkIterator(c, it, expected)
+	checkIterator(c, it, expected)
```
Comment (lamxTyler) [inline util/chunk/iterator_test.go]: Check it twice?

File: util/chunk/iterator_test.go (hunk 20 above)
Comment (lamxTyler) [reply util/chunk/iterator_test.go]: Can we extract these as a function? It occurs so many times.

File: executor/index_lookup_join.go [hunk 22]
```diff
@@ -177,30 +178,16 @@ func (e *IndexLookUpJoin) newInnerWorker(taskCh chan *lookUpJoinTask) *innerWork
...
+	return errors.Trace(err)
 }
 
 // Next implements the Executor interface.
 // Even though we only support read children in chunk mode, but we are not sure if our parents
 // support chunk, so we have to implement Next.
 func (e *IndexLookUpJoin) Next(goCtx goctx.Context) (Row, error) {
 	for {
-		err := e.prepareJoinResult(goCtx)
+		err := e.prepareJoinResult(goCtx, e.joinResult, false)
```
Comment (coocood) [inline executor/index_lookup_join.go]: We should add tests for `Next` mode for this Executor.

File: util/chunk/iterator_test.go (hunk 21 above)
Comment (XuHuaiyu) [reply util/chunk/iterator_test.go]: The loop can not be extracted as a function,
since the last parameter of c.Assert() is not the same type.

**Comment by XuHuaiyu** [comment]
PTAL @coocood

File: executor/join_test.go [hunk 23]
```diff
@@ -767,3 +768,30 @@ func (s *testSuite) TestIssue5278(c *C) {
...
+
+func (s *testSuite) TestIndexLookupJoin(c *C) {
+	cfg := config.GetGlobalConfig()
+	oldEnableChunk := cfg.EnableChunk
+	defer func() {
+		cfg.EnableChunk = oldEnableChunk
+	}()
+	tk := testkit.NewTestKit(c, s.store)
+	for _, enableChk := range []bool{true, false} {
+		cfg.EnableChunk = enableChk
```
Comment (coocood) [inline executor/join_test.go]: We don't need to set GlobalConfig, set session is enough.

**Comment by XuHuaiyu** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Comment by ngaut** [comment]
Amazing.

**Pull request merged**